client/network/src/service.rs

```diff
@@ -371,7 +371,19 @@ where
 				}
 			};
 
-			let mut builder = SwarmBuilder::new(transport, behaviour, local_peer_id)
+			let builder = if let Some(spawner) = params.executor {
+				struct SpawnImpl<F>(F);
+				impl<F: Fn(Pin<Box<dyn Future<Output = ()> + Send>>)> Executor for SpawnImpl<F> {
+					fn exec(&self, f: Pin<Box<dyn Future<Output = ()> + Send>>) {
+						(self.0)(f)
+					}
+				}
+				SwarmBuilder::with_executor(transport, behaviour, local_peer_id, SpawnImpl(spawner))
+			} else {
+				let tp = ThreadPoolBuilder::new().name_prefix("libp2p-swarm-task-").create()?;
+				SwarmBuilder::with_executor(transport, behaviour, local_peer_id, tp)
+			};
```

Comment:
This now causes a merge conflict with https://github.com/paritytech/substrate/pull/12646...